pkg/custompluginmonitor/types/config.go

```
@@ -0,0 +1,133 @@
+	if cpc.PluginGlobalConfig.TimeoutString == nil {
+		cpc.PluginGlobalConfig.TimeoutString = &defaultGlobalTimeoutString
+	}
+
+	timeout, err := time.ParseDuration(*cpc.PluginGlobalConfig.TimeoutString)
+	if err != nil {
+		return fmt.Errorf("Error in parsing global timeout %q. %v", *cpc.PluginGlobalConfig.TimeoutString, err)
+	}
+
+	cpc.PluginGlobalConfig.Timeout = &timeout
+
+	if cpc.PluginGlobalConfig.InvokeIntervalString == nil {
+		cpc.PluginGlobalConfig.InvokeIntervalString = &defaultInvokeIntervalString
+	}
+
+	invoke_interval, err := time.ParseDuration(*cpc.PluginGlobalConfig.InvokeIntervalString)
+	if err != nil {
+		return fmt.Errorf("Error in parsing invoke interval %q. %v", *cpc.PluginGlobalConfig.InvokeIntervalString, err)
+	}
+
+	cpc.PluginGlobalConfig.InvokeInterval = &invoke_interval
+
+	if cpc.PluginGlobalConfig.MaxOutputLength == nil {
+		cpc.PluginGlobalConfig.MaxOutputLength = &defaultMaxOutputLength
+	}
+	if cpc.PluginGlobalConfig.Concurrency == nil {
+		cpc.PluginGlobalConfig.Concurrency = &defaultConcurrency
+	}
+
+	for _, rule := range cpc.Rules {
+		if rule.TimeoutString != nil {
+			timeout, err := time.ParseDuration(*rule.TimeoutString)
+			if err != nil {
+				return fmt.Errorf("Error in parsing rule timeout %+v. %v", rule, err)
+			}
+			rule.Timeout = &timeout
+		}
+	}
+
+	return nil
+}
+
+// Validate verifies whether the settings in CustomPluginConfig are valid.
+func (cpc CustomPluginConfig) Validate() error {
+	if cpc.Plugin != customPluginName {
+		return fmt.Errorf("NPD does not support %q plugin for now. Only support \"custom\"", cpc.Plugin)
+	}
+
+	for _, rule := range cpc.Rules {
+		if rule.Timeout != nil && *rule.Timeout > *cpc.PluginGlobalConfig.Timeout {
+			return fmt.Errorf("Plugin timeout is greater than global timeout. "+
+				"Rule: %+v. Global timeout: %v", rule, cpc.PluginGlobalConfig.Timeout)
+		}
+	}
+
+	for _, rule := range cpc.Rules {
+		if _, err := os.Stat(rule.Path); os.IsNotExist(err) {
```

We should return any error, right?

---

Actually we have returned an error to represent the file not exist error. I am probably not understanding this comment quite well. :(

---

I mean we should return any error returned by `Stat`, right?